File: JeongSH/Programmers/DP/Triangle.java
/*
    7                (n=0층)
    3 8              (n=1층)
    8 1 0            (n=2층)
    2 7 4 4           ...
    4 5 2 6 5          .

    n=1층의 삼각형으로 생각했을 때,
    각각의 삼각형 갈림길의 기댓값은
    7
    (7+3) (7+8) 가됩니다.

    n=2층의 삼각형으로 생각했을 때,
    7
    (7+3) (7+8)
    (7+3+8) (Max((7+3),(7+8))+1) (7+8) 가 됩니다.

    시간복잡도: T(n) = n^2 = O(n^2)

 */
public class Triangle {
    public static void main(String args[]) {
        int[][] triangle = {{7}, {3, 8}, {8, 1, 0}, {2, 7, 4, 4}, {4, 5, 2, 6, 5}};
        int answer = 0;

        for (int i = 1; i < triangle.length; i++) {
            triangle[i][0] += triangle[i - 1][0];
            triangle[i][i] += triangle[i - 1][i - 1];
            for (int j = 1; j < i; j++) {
                triangle[i][j] += Math.max(triangle[i - 1][j - 1], triangle[i - 1][j]);
                answer = Math.max(triangle[i][j], answer);
            }
        }
        System.out.println(answer);
       return;
    }
}

File: 6loss0m/Programmers/DP/triangle.py
def solution(triangle):
    for j in range(1, len(triangle)):
        for i in range(j+1):
            # 삼각형 각층의 첫번째인 경우
            if i==0:
                triangle[j][0] += triangle[j-1][0]
            # 삼각형 각층의 마지막인 경우
            elif i==j:
                triangle[j][-1] += triangle [j-1][-1]
            # 내부 부분인 경우
            else:
                triangle[j][i] += max(triangle[j-1][i-1],triangle[j-1][i])
    answer = max(triangle[-1])
    return answer

# 참고 : https://codedrive.tistory.com/49